hadml/models/cgan/multihadron_event_gan.py

```diff
@@ -18,52 +18,92 @@ def __init__(
+        if loss_type == "wasserstein":
+            loss_gen = -score.mean(0).view(1)
+        elif loss_type == "bce":
+            loss_gen = torch.nn.functional.binary_cross_entropy_with_logits(
+                score, torch.ones_like(score))
+        elif loss_type == "ls":
+            loss_gen = 0.5 * ((score - 1) ** 2).mean(0).view(1)
+        return loss_gen
 
+    def _discriminator_loss(self, score_for_real, score_for_fake):
+        loss_type = self.hparams.loss_type
+        if loss_type == "wasserstein":
+            loss_disc = score_for_fake.mean(0).view(1) - score_for_real.mean(0).view(1)
+        elif loss_type == "bce":
+            loss_disc = torch.nn.functional.binary_cross_entropy_with_logits(
+                score_for_real, torch.ones_like(score_for_real)) + \
+                torch.nn.functional.binary_cross_entropy_with_logits(
+                    score_for_fake, torch.zeros_like(score_for_fake))
+        elif loss_type == "ls":
+            loss_disc = 0.5 * ((score_for_real - 1)**2).mean(0).view(1) + \
+                0.5 * (score_for_fake**2).mean(0).view(1)
+        return loss_disc
+    
     def validation_step(self, batch, batch_idx):
         gen_input, disc_input = batch
-        device = gen_input.device
         if self.trainer.state.stage == "validate":
-            batch_size = len(disc_input)
-            n_types = len(disc_input[0, 0, 4:])
-            max_n_hadrons = len(disc_input[0])
-            true_types = torch.stack([torch.argmax(d[:, 4:], dim=1) for d in disc_input])
-            # ==========================================================================================
-            # Simulating model predictions on validation data by providing some noise to the true values
-            # ==========================================================================================
-            distortion_noise =  torch.randint(0, n_types, (batch_size, max_n_hadrons))
-            distorted_types = (true_types + distortion_noise.to(device)) % n_types
-            distorted_disc_input = torch.stack([
-                torch.column_stack([
-                    d[:, :4] * torch.rand(1).to(device),           # scaling kinematics (Gaussian noise)
-                    torch.nn.functional.one_hot(d_type,            # randomly shifted types
-                                                num_classes=n_types)            
-                ]) for d, d_type in zip(disc_input, distorted_types)
-            ])
-            # ==========================================================================================
-            return {"distorted_disc_input": distorted_disc_input, "disc_input": disc_input}
-        
+            noise = self._generate_noise(*gen_input.size()[:2])
+            gen_output = self.generator(noise.to(gen_input.device), gen_input).detach()
+            return {"gen_output": gen_output, "disc_input": disc_input.detach()}
         elif self.trainer.state.stage == "sanity_check":
```

Comment:
Does the validation step break during the sanity check? If it doesn't, I would remove this if-else statement. Otherwise, this is good as a fix.